### PR TITLE
Add unlink to reference model tests

### PR DESCRIPTION
**Pull request by jamesbornholt**
## Description of change

This is pretty annoying because of some weird edge cases around implicit
directories being removed while local children are present. I think in
the long term we want to fix this, but it's a similar problem to what we
saw in #359 -- we need `readdir` to clean up removed directories
properly. So for now, I've changed the reference model to match our
current semantics, which is that if an ancestor of a local
file/directory is removed, that file/directory will no longer be visible
through the filesystem. Of course, once that file/directory becomes
remote it will become visible, and the model still captures that.

Relevant issues: #78 

## Does this change impact existing behavior?

Yes, but indirectly. We've changed the reference model to allow a different set of behavior around implicit directories, so bugs that the tests might have caught before could now be missed, and vice versa.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/tests/reftests/generators.rs [hunk 1]
```diff
@@ -120,7 +120,16 @@ pub fn flatten_tree(node: TreeNode) -> Vec<(String, MockObject)> {
...
+    // We allow names in the tree to contain `/`, which can cause duplicate keys
+    let mut keys = HashSet::new();
+    let mut ret = Vec::with_capacity(contents.len());
+    for (key, object) in contents {
+        if !keys.contains(&key) {
+            keys.insert(key.clone());
+            ret.push((key, object));
+        }
+    }
     ret
```
[the remark this answers is not in the document]
Comment (jamesbornholt) [reply mountpoint-s3/tests/reftests/generators.rs:134]: sneaky, I like it

**Review by passaro — APPROVED** [review]
